runtime/src/main/java/org/corfudb/runtime/collections/TxnContext.java

```diff
@@ -0,0 +1,431 @@
+     * @param key   Key of the record.
+     * @param value Value or payload of the record.
+     * @param metadata Metadata associated with the record.
+     * @param <K>   Type of Key.
+     * @param <V>   Type of Value.
+     * @param <M>   Type of Metadata.
+     * @return TxnContext instance.
+     */
+    @Nonnull
+    public <K extends Message, V extends Message, M extends Message>
+    TxnContext put(@Nonnull Table<K, V, M> table,
+                   @Nonnull final K key,
+                   @Nonnull final V value,
+                   @Nullable final M metadata) {
+        validateTableWrittenIsInNamespace(table);
+        operations.add(() -> {
+            table.put(key, value, metadata);
+            table.getMetrics().numPuts.incrementAndGet();
+        });
+        return this;
+    }
+
+    /**
+     * Merges the delta value with the old value by applying a caller specified BiFunction and writes
+     * the final value.
+     *
+     * @param table Table object to perform the merge operation on.
+     * @param key            Key
+     * @param mergeOperator  Function to apply to get the new value
+     * @param recordDelta    Argument to pass to the mutation function
+     * @param <K>            Type of Key.
+     * @param <V>            Type of Value.
+     * @param <M>            Type of Metadata.
+     * @return TxnContext instance
+     */
+    @Nonnull
+    public <K extends Message, V extends Message, M extends Message>
+    TxnContext merge(@Nonnull Table<K, V, M> table,
+                     @Nonnull final K key,
+                     @Nonnull BiFunction<CorfuRecord<V, M>, CorfuRecord<V,M>, CorfuRecord<V,M>> mergeOperator,
+                     @Nonnull final CorfuRecord<V,M> recordDelta) {
+        validateTableWrittenIsInNamespace(table);
+        operations.add(() -> {
+            CorfuRecord<V,M> oldRecord = table.get(key);
+            CorfuRecord<V, M> mergedRecord;
+            try {
+                mergedRecord = mergeOperator.apply(oldRecord, recordDelta);
+            } catch (Exception ex) {
+                txAbort(); // explicitly abort this transaction and then throw the abort manually
+                log.error("TX Abort merge: {}", table.getFullyQualifiedTableName(), ex);
+                throw ex;
+            }
+            table.put(key, mergedRecord.getPayload(), mergedRecord.getMetadata());
+            table.getMetrics().numMerges.incrementAndGet();
+        });
+        return this;
+    }
+
+    /**
+     * touch() is a call to create a conflict on a read in a write-only transaction
+     *
+     * @param table Table object to perform the create/update on.
+     * @param key   Key of the record to touch.
+     * @param <K>   Type of Key.
+     * @param <V>   Type of Value.
+     * @param <M>   Type of Metadata.
+     * @return TxnContext instance.
+     */
+    @Nonnull
+    public <K extends Message, V extends Message, M extends Message>
+    TxnContext touch(@Nonnull Table<K, V, M> table,
```

Comment:
Touch seems to be a read & write operation, so the txnType needs to be READ_WRITE?

Reply:
no, touch is entirely encapsulated within the lambda so we treat it as a write-only operation.